JitsiAuthConnection.js

```diff
@@ -0,0 +1,162 @@
+     *
+     * NOTE If neither one of the errors is present, then the operation has been
+     * canceled.
+     */
+
+    /**
+     * Connects to the XMPP server using the specified credentials and contacts
+     * Jicofo in order to obtain a session ID (which is then stored in the local
+     * storage). The user's role of the parent conference will be upgraded to
+     * moderator (by Jicofo).
+     *
+     * @param {Object} options
+     * @param {string} options.id - XMPP user's ID to log in. For example,
+     * user@xmpp-server.com.
+     * @param {string} options.password - XMPP user's password to log in with.
+     * @param {string} [options.roomPassword] - The password to join the MUC
+     * with.
+     * @param {Function} [options.onLoginSuccessful] - Callback called when
+     * logging into the XMPP server was successful. The next step will be to
+     * obtain a new session ID from Jicofo and join the MUC using it which will
+     * effectively upgrade the user's role to moderator.
+     * @returns {Promise} Resolved in case the authentication was successful
+     * and the session ID has been stored in the settings. Will be rejected with
+     * {@link UpgradeRoleError} which will have either <tt>connectionError</tt>
+     * or <tt>authenticationError</tt> property set depending on which of the
+     * steps has failed. If {@link cancel} is called before the operation is
+     * finished, then the promise will be rejected with an empty object (i.e. no
+     * error set).
+     */
+    authenticateAndUpgradeRole({
+        // 1. Log the specified XMPP user in.
+        id,
+        password,
+
+        // 2. Let the API client/consumer know as soon as the XMPP user has been
+        //    successfully logged in.
+        onLoginSuccessful,
+
+        // 3. Join the MUC.
+        roomPassword
+    }) {
+        if (this._promise) {
+            return this._promise;
+        }
+
+        this._promise = new Promise((resolve, reject) => {
+            const connectionEstablished = () => {
+                if (this.canceled) {
+                    return;
+                }
+
+                // Let the caller know that the XMPP login was successful.
+                onLoginSuccessful && onLoginSuccessful();
+
+                // Now authenticate with Jicofo and get a new session ID.
+                this.room
+                    = this.xmpp.createRoom(
+                        this.conference.options.name,
+                        this.conference.options.config);
+                this.room.moderator.authenticate()
+                    .then(() => {
+                        if (this.canceled) {
+                            return;
+                        }
+
+                        this.xmpp.disconnect();
+
+                        // At this point we should have the new session ID
+                        // stored in the settings. Jicofo will allow to join the
+                        // room.
```

Comment:
* @saghul asked where that's "stored in the settings". I think that's `this.room.moderator.authenticate` which does `Settings.setSessionId`.

* I'm not really sure why `sessionId` is stored in `localStorage` given that `localStorage` is expected to (maybe) survive across browser sessions and the piece of data suggests that it's a session-related piece of data.